dags/synaptor_ops.py

```diff
@@ -21,13 +21,145 @@
 TASK_QUEUE_NAME = "synaptor"
 
 
-# Python callables (for PythonOperators)
-def generate_ngl_link_op() -> None:
+# Op functions
+def generate_ngl_link(
+    net_output_path: str,
+    seg_path: str,
+    workflowtype: str,
+    storagedir: str,
+    img_path: Optional[str] = None,
+    voxelres: Optional[tuple[int, int, int]] = None,
+) -> None:
     """Generates a neuroglancer link to view the results."""
-    pass
+    layers = [
+        ImageLayer("network output", net_output_path),
+        SegLayer("synaptor segmentation", seg_path),
+    ]
+
+    if img_path:
+        layers = [ImageLayer("image", img_path)] + layers
+
+    payload = generate_ng_payload(layers)
+
+    if "Assignment" in workflowtype:
+        presyn_pts, postsyn_pts = read_pts(storagedir)
+        payload = add_annotation_layer(payload, presyn_pts, postsyn_pts, voxelres)
+
+    basename = dump_payload(payload, storagedir)
+
+    slack_message(wrap_payload(os.path.join(storagedir, basename)), broadcast=True)
+
+
+def dump_payload(payload: dict, storagedir: str) -> str:
+    from cloudfiles import CloudFiles
+    basename = "ng.json"
+
+    cf = CloudFiles(storagedir)
+    cf.put_json(basename, payload)
+
+    return basename
+
+
+def read_pts(storagedir: str) -> tuple[list, list]:
+    from cloudfiles import CloudFiles
+
+    cf = CloudFiles(storagedir)
+    maybe_content = cf.get("final_edgelist.df")
+
+    if maybe_content:
+        content = maybe_content.decode("utf-8")
+    else:
+        raise ValueError("no edge list found")
+
+    lines = content.split("\n")
+    header, rows = lines[0], lines[1:]
+
+    # indices for the columns we want
+    colnames = header.split(",")
+    pre_x_i = colnames.index("presyn_x")
+    pre_y_i = colnames.index("presyn_y")
+    pre_z_i = colnames.index("presyn_z")
+    post_x_i = colnames.index("postsyn_x")
+    post_y_i = colnames.index("postsyn_y")
+    post_z_i = colnames.index("postsyn_z")
+
+    # extracting points
+    presyn_pts = list()
+    postsyn_pts = list()
+    for row in rows:
+        if "," not in row:
+            continue
+
+        fields = row.split(",")
+        presyn_pt = fields[pre_x_i], fields[pre_y_i], fields[pre_z_i]
+        postsyn_pt = fields[post_x_i], fields[post_y_i], fields[post_z_i]
+
+        presyn_pts.append(list(map(int, presyn_pt)))
+        postsyn_pts.append(list(map(int, postsyn_pt)))
+
+    return presyn_pts, postsyn_pts
+
+
+def add_annotation_layer(
+    payload: dict, presyn_pts: list, postsyn_pts: list, voxel_res: tuple
+) -> dict:
+    viewer_res = payload["navigation"]["pose"]["position"]["voxelSize"]
+    factors = [vox_res / view_res for view_res, vox_res in zip(viewer_res, voxel_res)]
```

Comment:
I think annotation layers can specify their own `voxelSize` , do you still need to scale everything to match the resolution of the viewer?